lib/client/snapshot-request.js

```diff
@@ -0,0 +1,78 @@
+var hat = require('hat');
+
+module.exports = SnapshotRequest;
+
+function SnapshotRequest(connection, collection, id, version, callback) {
+  this.requestId = hat();
+
+  this.connection = connection;
+  this.id = id;
+  this.collection = collection;
+  this.callback = callback;
+
+  this.parseVersion(version);
+
+  this.ready = false;
+  this.sent = false;
+}
+
+SnapshotRequest.prototype.parseVersion = function (version) {
+  if (typeof version === 'number') {
+    this.version = version;
+  } else if (version instanceof Date) {
+    this.timestamp = version.getTime();
+  } else if (!version) {
+    this.version = null;
+  } else {
+    this.callback({ code: 4024, message: 'Invalid version' });
+  }
+}
+
+SnapshotRequest.prototype.hasPending = function () {
+  return !this.ready;
+};
+
+SnapshotRequest.prototype.send = function () {
+  if (!this.connection.canSend) {
+    return;
+  }
+
+  var message = {
+    a: 'sv',
+    id: this.requestId,
+    c: this.collection,
+    d: this.id,
+    v: this.version,
+    ts: this.timestamp
+  };
+
+  this.connection.send(message);
+  this.sent = true;
+};
+
+SnapshotRequest.prototype._onConnectionStateChanged = function () {
+  if (this.connection.canSend && !this.sent) {
+    this.send();
+  }
```

Comment:
If `!this.connection.canSend`, then probably set `this.sent` to `false`, so that the request would be resent after re-connection.

Reply:
It starts as `false`, and is only changed to `true` at the end of `send()`.

Reply:
Aha cool. Still trying to piece together all the data flows around this library! Some of it's pretty confusing.